### PR TITLE
fix: adding settings that are missing

### PR DESCRIPTION
**Pull request by MWClayson-NHS**
## Description

Adding validation rules to prevent invalid records from processing 

## Context

ensuring bad data doesn't get sent to the cohort distribution. 

## Type of changes

- [ ] Refactoring (non-breaking change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [x] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [x] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: .vscode/tasks.json
Comment (andrewmorris43) [inline .vscode/tasks.json:1]: Why are all these functions being removed from this file?

File: application/CohortManager/src/Functions/CaasIntegration/processCaasFile/processCaasFile.cs [hunk 1]
```diff
@@ -8,6 +8,7 @@ namespace NHS.CohortManager.CaasIntegrationService;
 using Microsoft.Extensions.Logging;
 using Microsoft.Azure.Functions.Worker;
 using Microsoft.Azure.Functions.Worker.Http;
+using CsvHelper;
```
Comment (Warren-Pitterson) [inline application/CohortManager/src/Functions/CaasIntegration/processCaasFile/processCaasFile.cs:11]: Is this used in this class?

File: application/CohortManager/src/Functions/CaasIntegration/processCaasFile/processCaasFile.cs [hunk 2]
```diff
@@ -104,20 +105,36 @@ public async Task<HttpResponseData> Run([HttpTrigger(AuthorizationLevel.Anonymou
...
-                        await _callFunction.SendPost(Environment.GetEnvironmentVariable("StaticValidationURL"), json);
-                        _logger.LogInformation("Called static validation");
+                            });
+                            return _createResponse.CreateHttpResponse(HttpStatusCode.BadRequest, req);
+                        }
                     }
                     catch (Exception ex)
                     {
-                        _logger.LogError("Static validation function failed.\nMessage: {Message}\nStack Trace: {StackTrace}", ex.Message, ex.StackTrace);
+                        _logger.LogError("handling the exception failed.\nMessage: {Message}\nStack Trace: {StackTrace}", ex.Message, ex.StackTrace);
```
Comment (Warren-Pitterson) [inline application/CohortManager/src/Functions/CaasIntegration/processCaasFile/processCaasFile.cs:137]: Capital please, should this be "Handling"

File: application/CohortManager/src/Functions/CaasIntegration/processCaasFile/processCaasFile.cs [hunk 3]
```diff
@@ -133,4 +150,16 @@ public async Task<HttpResponseData> Run([HttpTrigger(AuthorizationLevel.Anonymou
 
         return _createResponse.CreateHttpResponse(HttpStatusCode.OK, req);
     }
+
+    private async Task<bool> PostDemographicDataAsync(Participant participant)
+    {
+        var demographicDataInserted = await _checkDemographic.PostDemographicDataAsync(participant, Environment.GetEnvironmentVariable("DemographicURI"));
+        if (demographicDataInserted == false)
```
Comment (Warren-Pitterson) [inline application/CohortManager/src/Functions/CaasIntegration/processCaasFile/processCaasFile.cs:157]: Can this be !demographicDataInserted instead of == false

File: application/CohortManager/src/Functions/CohortDistributionServices/ValidateCohortDistributionRecord/host.json
Comment (andrewmorris43) [inline application/CohortManager/src/Functions/CohortDistributionServices/ValidateCohortDistributionRecord/host.json:1]: What is this file used for?

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs [hunk 4]
```diff
@@ -140,11 +140,12 @@ private async Task InsertValidationErrorIntoDatabase(string fileName)
         });
 
         var result = await _callFunction.SendPost(Environment.GetEnvironmentVariable("FileValidationURL"), json);
-        if (result.StatusCode == HttpStatusCode.OK)
+        if (result.StatusCode != HttpStatusCode.OK)
         {
-            _logger.LogInformation("file failed checks and has been moved to the poison blob storage");
+            _logger.LogError("there was a problem saving and or moving the failed file");
```
Comment (Warren-Pitterson) [inline application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs:145]: can we include the filename please and maybe reword the message to 

"An error occurred while saving or moving the failed file {filename}."

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs [hunk 5]
```diff
@@ -140,11 +140,12 @@ private async Task InsertValidationErrorIntoDatabase(string fileName)
...
         }
-        _logger.LogError("there was a problem saving and or moving the failed file");
+        _logger.LogInformation("file failed checks and has been moved to the poison blob storage");
```
Comment (Warren-Pitterson) [inline application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs:147]: Capital F on file please. Can we also include the filename

File: application/CohortManager/src/dotnet-mesh-client
Comment (andrewmorris43) [inline application/CohortManager/src/dotnet-mesh-client:1]: Do we need this file?

File: application/CohortManager/src/Functions/Shared/Model/Participant.cs [hunk 6]
```diff
@@ -4,7 +4,7 @@ namespace Model;
 
 public class Participant
 {
-    public Participant() {}
+    public Participant() { }
```
Comment (Warren-Pitterson) [inline application/CohortManager/src/Functions/Shared/Model/Participant.cs:7]: Was this space added intentionally?

File: tests/CaasIntegrationTests/processCaasFileTest/processCaasFileTests.cs [hunk 7]
```diff
@@ -111,18 +113,19 @@ public async Task Run_Should_Call_RemoveParticipant_For_Each_Removed_Participant
         _callFunction.Verify(
             x => x.SendPost(It.Is<string>(s => s.Contains("PMSRemoveParticipant")), It.IsAny<string>()),
             Times.Exactly(2));
+
         _callFunction.VerifyNoOtherCalls();
     }
 
     [TestMethod]
-    public async Task Run_Should_Call_StaticValidation_For_Each_Unknown_Participant_RecordType()
+    public async Task Run_Should_Create_new_exception_For_Each_Unknown_Participant_RecordType()
```
Comment (Warren-Pitterson) [inline tests/CaasIntegrationTests/processCaasFileTest/processCaasFileTests.cs:121]: Can the test follow the naming method of MethodName_WhatIsBeingTest_ExpectedResult please

File: application/CohortManager/src/Functions/Shared/Common/ExceptionHandler.cs [hunk 8]
```diff
@@ -96,6 +96,36 @@ public async Task<bool> CreateValidationExceptionLog(IEnumerable<RuleResultTree>
...
+            ExceptionDate = validation.ExceptionDate ?? DateTime.Now,
+            RuleContent = validation.RuleContent ?? "",
+            ScreeningService = validation.ScreeningService ?? 0
+        };
+
+        var url = GetUrlFromEnvironment();
+        var response = await _callFunction.SendPost(url, JsonSerializer.Serialize(requestObject));
+        if (response.StatusCode != HttpStatusCode.OK)
+        {
+            _logger.LogError("there was an error while logging an exception to the database");
```
Comment (Warren-Pitterson) [inline application/CohortManager/src/Functions/Shared/Common/ExceptionHandler.cs:123]: Capital T please

File: application/CohortManager/src/Functions/ScreeningValidationService/FileValidation/FileValidation.cs [hunk 9]
```diff
@@ -44,10 +44,14 @@ public async Task<HttpResponseData> RunAsync([HttpTrigger(AuthorizationLevel.Ano
                 DateCreated = requestBody.DateCreated ?? DateTime.Now,
                 FileName = string.IsNullOrEmpty(requestBody.FileName) ? "" : requestBody.FileName,
                 DateResolved = requestBody.DateResolved ?? DateTime.MaxValue,
-                RuleDescription = requestBody.RuleDescription ?? "",
+                RuleDescription = requestBody.RuleDescription ?? "the file failed file validation. Check the file Exceptions blob store ",
```
Comment (Warren-Pitterson) [inline application/CohortManager/src/Functions/ScreeningValidationService/FileValidation/FileValidation.cs:47]: Capital T on the

File: infrastructure/modules/function-app/locals.tf [hunk 10]
```diff
@@ -54,39 +55,42 @@ locals {
...
     }
 
     CreateParticipant = {
 
-      DtOsDatabaseConnectionString = local.db_connection_string
-      LookupValidationURL          = local.fnapp_urls.lookupValidation
-      ExceptionFunctionURL         = local.fnapp_urls.createException
+      DtOsDatabaseConnectionString             = local.db_connection_string
+      LookupValidationURL                      = local.fnapp_urls.lookupValidation
+      ExceptionFExceptionFunctionURLunctionURL = local.fnapp_urls.createException
```
Comment (rfk-nc) [inline infrastructure/modules/function-app/locals.tf:93]: Typo in name: ExceptionFExceptionFunctionURLunctionURL

File: application/CohortManager/src/dotnet-mesh-client
Comment (MWClayson-NHS) [reply application/CohortManager/src/dotnet-mesh-client:1]: This came from my local branch, it will be a submodule once i merge my changes in.

File: application/CohortManager/src/Functions/Shared/Common/ExceptionHandler.cs [hunk 11]
```diff
@@ -96,6 +96,36 @@ public async Task<bool> CreateValidationExceptionLog(IEnumerable<RuleResultTree>
...
+    {
+        var requestObject = new ValidationException()
+        {
+            RuleId = validation.RuleId == null ? 0 : validation.RuleId,
+            Cohort = "",
+            NhsNumber = string.IsNullOrEmpty(validation.NhsNumber) ? "" : validation.NhsNumber,
+            DateCreated = validation.DateCreated ?? DateTime.Now,
+            FileName = string.IsNullOrEmpty(validation.FileName) ? "" : validation.FileName,
+            DateResolved = validation.DateResolved ?? DateTime.MaxValue,
+            RuleDescription = validation.RuleDescription ?? "the file failed file validation. for single record",
```
Comment (stephhou) [inline application/CohortManager/src/Functions/Shared/Common/ExceptionHandler.cs:109]: This rule description default needs a bit of work.

I think this might also have been fixed in Sam's recent ticket.

File: application/CohortManager/src/Functions/Shared/Common/ExceptionHandler.cs [hunk 12]
```diff
@@ -96,6 +96,36 @@ public async Task<bool> CreateValidationExceptionLog(IEnumerable<RuleResultTree>
...
+            Category = validation.Category ?? 0,
+            ScreeningName = validation.ScreeningName ?? "",
+            Fatal = validation.Fatal ?? 1,
+            ErrorRecord = validation.ErrorRecord ?? "",
+            ExceptionDate = validation.ExceptionDate ?? DateTime.Now,
+            RuleContent = validation.RuleContent ?? "",
```
Comment (stephhou) [inline application/CohortManager/src/Functions/Shared/Common/ExceptionHandler.cs:115]: ErrorRecord replaced RuleContent. I can raise a ticket for checking if it is still being used in other areas of the app

File: application/CohortManager/src/Functions/Shared/Common/ExceptionHandler.cs [hunk 13]
```diff
@@ -96,6 +96,36 @@ public async Task<bool> CreateValidationExceptionLog(IEnumerable<RuleResultTree>
...
+            ScreeningService = validation.ScreeningService ?? 0
```
Comment (stephhou) [inline application/CohortManager/src/Functions/Shared/Common/ExceptionHandler.cs:116]: The exception table doesn't have ScreeningService as a field, are we using that for something else?

File: application/CohortManager/src/Functions/Shared/Model/Constants/Actions.cs [hunk 14]
```diff
@@ -1,8 +1,38 @@
...
+public class Actions
 {
     public const string New = "New";
     public const string Amended = "Amended";
     public const string Removed = "Removed";
+
+
+    /// <summary>
+    /// gets all the actions and tries to parse the Given action to the available action
+    /// </summary>
```
Comment (stephhou) [inline application/CohortManager/src/Functions/Shared/Model/Constants/Actions.cs:14]: The action names are changing to ADD, AMENDED and DELETED. The record will be refused if it doesn't match exactly (including case) in the switch statement in ProcessCaasFile. Where has this change come from?

File: application/CohortManager/src/Functions/CaasIntegration/processCaasFile/processCaasFile.cs [hunk 15]
```diff
@@ -104,20 +105,36 @@ public async Task<HttpResponseData> Run([HttpTrigger(AuthorizationLevel.Anonymou
...
+                            await _handleException.CreateRecordValidationExceptionLog(new Model.ValidationException()
+                            {
+                                RuleId = 1,
+                                Cohort = "N/A",
+                                NhsNumber = string.IsNullOrEmpty(participant.NhsNumber) ? "" : participant.NhsNumber,
+                                DateCreated = DateTime.Now,
+                                FileName = string.IsNullOrEmpty(basicParticipantCsvRecord.FileName) ? "" : basicParticipantCsvRecord.FileName,
+                                DateResolved = DateTime.MaxValue,
+                                RuleDescription = $"a record has failed to process with the NHS Number: {participant.NhsNumber} because the of an incorrect record type",
+                                Category = 0,
```
Comment (stephhou) [inline application/CohortManager/src/Functions/CaasIntegration/processCaasFile/processCaasFile.cs:123]: Should be category 1

File: application/CohortManager/src/Functions/CaasIntegration/processCaasFile/processCaasFile.cs [hunk 16]
```diff
@@ -104,20 +105,36 @@ public async Task<HttpResponseData> Run([HttpTrigger(AuthorizationLevel.Anonymou
...
                         {
-                            FileName = input.FileName,
-                            Participant = participant
-                        };
+                            _logger.LogError($"cannot parse record type with action: {participant.RecordType}");
+
+                            await _handleException.CreateRecordValidationExceptionLog(new Model.ValidationException()
+                            {
+                                RuleId = 1,
+                                Cohort = "N/A",
```
Comment (stephhou) [inline application/CohortManager/src/Functions/CaasIntegration/processCaasFile/processCaasFile.cs:117]: We should have the Cohort name, it might be in a ticket to be merged after this one. If so I'll make another ticket to address it later

Comment (SamRobinson75684) [reply application/CohortManager/src/Functions/CaasIntegration/processCaasFile/processCaasFile.cs:117]: I think this in an another thing to be merged aftrerwards yes
